velox/expression/CastExpr-inl.h

```diff
@@ -46,8 +46,175 @@ inline std::exception_ptr makeBadCastException(
+  if (s[pos] == '-') {
+    out.sign = -1;
+    ++pos;
+  } else if (s[pos] == '+') {
+    out.sign = 1;
+    ++pos;
+  }
+  // First run of digits.
+  pos = parseDigitsRun(s, pos, size, out.wholeDigits);
+  if (pos == size) {
+    return out.wholeDigits.empty() ? std::nullopt
+                                   : std::optional<DecimalComponents>(out);
+  }
+  // Optional dot (if given in fractional form).
+  if (s[pos] == '.') {
+    // Second run of digits.
+    ++pos;
+    pos = parseDigitsRun(s, pos, size, out.fractionalDigits);
+  }
+  if (out.wholeDigits.empty() && out.fractionalDigits.empty()) {
+    // Need at least some digits (whole or fractional).
+    return std::nullopt;
+  }
+  if (pos == size) {
+    return out;
+  }
+  // Optional exponent.
+  if (s[pos] == 'e' || s[pos] == 'E') {
+    ++pos;
+    if (pos != size && s[pos] == '+') {
+      ++pos;
+    }
+    folly::StringPiece p = {s + pos, size - pos};
+    auto tryExp =
+        folly::tryTo<int32_t>(folly::StringPiece(s + pos, size - pos));
+    if (tryExp.hasError()) {
+      return std::nullopt;
+    }
+    out.exponent = tryExp.value();
+    return out;
+  }
+  return pos == size ? std::optional<DecimalComponents>(out) : std::nullopt;
+}
+
+/// Multiple out by the appropriate power of 10 necessary to add source parsed
+/// as int128_t and then adds the parsed value of source.
+bool shiftAndAdd(std::string_view input, int128_t& out) {
+  auto length = input.size();
+  if (length == 0) {
+    return true;
+  }
+  out = checkedMultiply<int128_t>(out, DecimalUtil::kPowersOfTen[length]);
+  auto tryValue =
+      folly::tryTo<int128_t>(folly::StringPiece(input.data(), length));
+  if (tryValue.hasError()) {
+    return false;
+  }
+  out = checkedPlus<int128_t>(tryValue.value(), out);
+  return true;
+}
+
+/// Derives from Arrow function DecimalFromString.
+/// Arrow implementation:
+/// https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/decimal.cc#L637
+///
+/// Firstly, it will parse the varchar to DecimalComponents which contains the
+/// message that can represent a value. Secondly, process the exponent to get
+/// the value parsedScale. Thirdly, compute the rescaled value.
+/// The caller should test if `error` is empty
+template <typename T>
+std::optional<T> rescaleVarchar(
+    const StringView s,
+    int toPrecision,
+    int toScale,
+    std::string& error) {
+  auto decimalComponentsOpt = parseDecimalComponents(s.data(), s.size());
+  if (!decimalComponentsOpt.has_value()) {
+    error = "Value is not a number.";
+    return std::nullopt;
+  }
+  auto decimalComponents = decimalComponentsOpt.value();
+
+  // Count number of significant digits (without leading zeros).
+  size_t firstNonZero = decimalComponents.wholeDigits.find_first_not_of('0');
+  size_t significantDigits = decimalComponents.fractionalDigits.size();
+  if (firstNonZero != std::string::npos) {
+    significantDigits += decimalComponents.wholeDigits.size() - firstNonZero;
```

Comment:
Let's update the documentation to include examples with multiple zeros at the start and end: '0000.123' and '.12300000000'